Extended VRM4U with Auto-Mapping and Performance Optimizations

desuqcafe opened the pull request:
# Extended VRM4U with Auto-Mapping and Performance Optimizations

## Update
- Rebased on VRM4U's latest commits which are going to support U.E. 5.6   ( 5/30/2025)

##  Auto-Mapping System
- **Smart skeleton detection** for VRM, Mixamo, MetaHuman, and DAZ character types
- **Automatic bone mapping** eliminates manual configuration for supported skeleton types
- **Custom bone overrides** support unconventional bone names and edge cases
- **Fallback detection** tries multiple naming conventions before failing

## ⚡ Performance Optimizations
### Resolved Critical Performance Issues:
- **Eliminated deep copying** of OSC messages (previously copying up to 90 times per second)
- **Removed thread blocking** during data processing
- **Fixed memory allocation issues** - stopped constant deallocation/allocation of large data structures
- **Optimized bone curve data** - eliminated repeated copying of bone curve data maps

### New Performance Profile System:
- **Performance Mode (30 FPS)** - Optimized for lower-end hardware
- **Balanced Mode (60 FPS)** - Good performance/quality balance
- **Streaming Mode (90 FPS)** - Maximum quality for live streaming
- **Adaptive Mode** - Automatically adjusts FPS based on system load
- **Custom Mode** - User-defined update rates (30-120 FPS)

## 🔧 Developer Experience Improvements
- **One-click Auto Populate** button for instant skeleton setup
- **Detailed logging** with mapping statistics and troubleshooting info
- **Graceful fallbacks** for unsupported skeleton types
- **Backward compatibility** with existing VRM workflows

## Technical Details
- Added `UAutoPopulateVrmMeta` utility class for skeleton detection
- Implemented `EVMCPerformanceMode` enum for structured performance control
- Extended `FAnimNode_VrmVMC` with intelligent throttling and auto-detection
- Comprehensive bone mapping for 4 major skeleton types with critical bone prioritization

**Impact:** Transforms VRM4U from VRM-specific tool into universal motion capture solution with significant performance gains and automated setup.

## Testing
- Verified in Unreal Engine 5.5 for all options.

MattiasOhmanSony committed:
Bugfix. Insufficient pattern matching can lead to crash.

by using continue the loop does not restart the pattern matching and it can lead to that we get faulty indexes that in turn can lead to crashing. The correct code here is to break in the inner loop so we return to the outer loop.

desuqcafe committed:
Add auto populate feature to VRMObject

Implements functionality to automatically populate VRM metadata, simplifying the avatar setup process.

desuqcafe committed:
Implement proper mapping of avatar bones

Define correct bone mapping structure to ensure accurate skeletal representation for Mixamo, MetaHuman, VRM.
Improves compatibility with standard VRM specifications.

desuqcafe committed:
Enhance auto population frontend and backend logic

Improve user interface controls and optimize underlying code for more efficient avatar bone mapping.
Add validation checks and error handling to make the feature more robust.

desuqcafe committed:
Update deprecated API calls for UE 5.5 compatibility

Replace outdated function calls and API usage to ensure compatibility with Unreal Engine 5.5.
Fix compiler warnings and potential errors caused by API changes.

desuqcafe committed:
Add support for DAZ similar to other skeletal mesh types

desuqcafe committed:
Fix VMC animation node array bounds crash

desuqcafe committed:
Optimize VMC performance with batching and add configurable performance modes

desuqcafe committed:
update: README.md

Replacing Ruyo's README with OWL's while also still maintaining Ruyo's README versions

desuqcafe committed:
add: README.md support files

campbellwmorgan commented on Source/VRM4UCapture/Private/VrmVMCObject.cpp:
```diff
@@ -35,72 +50,148 @@ void UVrmVMCObject::CreateServer(FString inName, uint16 inPort) {
+		} else if (FrameDelta < 1.0f / 60.0f) {
+			CurrentThrottleTime = 1.0f / 90.0f;  // System performing well
+		} else {
+			CurrentThrottleTime = UpdateThrottleTime; // Use base rate
```
Since you've used these a bit, elsewhere would it not make sense to store these fractions as constants somewhere?

desuqcafe replied:
Hmm, it's been awhile didn't see elsewhere I've used it nonetheless, refactored to make the code more readable for this part.

campbellwmorgan commented on Source/VRM4UCapture/Public/VrmVMCObject.h:
```diff
@@ -5,57 +5,73 @@
+       if (Port != Other.Port) return false;
+       if (ServerAddress != Other.ServerAddress) return false;
+       return true;
+    }
```
This has been converted from tabs to spaces. We should revert

campbellwmorgan commented on Source/VRM4UCaptureEditor/Private/AutoPopulateVrmMeta.cpp:
```diff
@@ -0,0 +1,1101 @@
+				break;
+			default: TypeName = TEXT("Unknown");
+			}
+			UE_LOG(LogTemp, Log, TEXT("Auto-detected skeleton type: %s"), *TypeName);
```
I think you're just converting this to string right? You can use `UEnum::GetValueAsString`

desuqcafe replied:
ah right :)

campbellwmorgan commented on Source/VRM4UCaptureEditor/Private/AutoPopulateVrmMeta.cpp:
```diff
@@ -0,0 +1,1101 @@
+				SkeletonType == ESkeletonType::VRM ? TEXT("VRM") :
+				SkeletonType == ESkeletonType::Mixamo ? TEXT("Mixamo") :
+				SkeletonType == ESkeletonType::MetaHuman ? TEXT("MetaHuman") :
+				SkeletonType == ESkeletonType::DAZ ? TEXT("DAZ") : TEXT("Unknown"));
```
can use `UEnum::GetValueAsString` here too

campbellwmorgan commented on Source/VRM4UCaptureEditor/VRM4UCaptureEditor.Build.cs:
```diff
@@ -9,30 +9,45 @@ public VRM4UCaptureEditor(ReadOnlyTargetRules Target) : base(Target)
+
+        DynamicallyLoadedModuleNames.AddRange(
+	        new string[] {
+	        });
```
Seem to be a tab / spaces thing going on here

campbellwmorgan commented on Source/VRM4ULoader/Private/VrmConvertMorphTarget.cpp:
```diff
@@ -367,34 +367,69 @@ bool VRMConverter::ConvertMorphTarget(UVrmAssetListObject *vrmAssetList) {
+       if (LODInfoPtr) {
+          LODInfoPtr->ReductionSettings.NumOfTrianglesPercentage = 1.f;
+       }
+    }
```
Spaces

campbellwmorgan reviewed (changes requested):
This is a great addition to the repo.

I only have some minor suggestions, however depending on what the plan is for this repo in relation to the Ruyo upstream I think we should:
1. Not modify the main readme except to provide a link to the wiki (hosted on our repo)
2. Move all the current OWL readme updates to the wiki on our repo
3. Not reformat any lines of code we have not modified

I suggest we do this before we merge to `master` on this repo so that we can then do a PR to the upstream repo without having to PR with revert commits etc...

ruyo committed:
ウインドウサイズ変更時にクラッシュするのを対処

ruyo committed:
CS整頓中

ruyo committed:
fix: build error ue5.3

ruyo committed:
Diffusionフィルタの端のノイズを修正

ruyo committed:
emissiveUV2追加

ruyo committed:
Merge pull request #446 from MattiasOhmanSony/master

Bugfix. Insufficient pattern matching can lead to crash.

ruyo committed:
JSONのデータが無い場合にクラッシュしてしまうのを修正

ruyo committed:
Shadeテクスチャが無い時、Mainを貼るよう修正

ruyo committed:
Valkan SM6でクラッシュするのを対処

ruyo committed:
RimLightFilter制御クラス追加

ruyo committed:
RimFilter クラッシュ対処

ruyo committed:
CS 整頓

ruyo committed:
CS 整頓

ruyo committed:
CS整頓

ruyo committed:
CS 整頓

ruyo committed:
マテリアルのプレフィクスを MIに変更

ruyo committed:
ステンシル無効時 CS無効化

ruyo committed:
fix: ビルドエラー

ruyo committed:
フィルタパラメータ 調整

ruyo committed:
Update README.md

ruyo committed:
VRMテクスチャ 範囲外をチェック

ruyo committed:
Update README.md

ruyo committed:
JSONをサイズから正しくパースできるようにした。

ruyo committed:
VRM1判定が狂うことがあるのを修正

ruyo committed:
assimp 更新

ruyo committed:
Update README.md

desuqcafe committed:
update: Cleanup code style, slight refactor

desuqcafe commented:
What's remaining will be the OWL README update, which I'll follow up with the person responsible since there have been some updates to it

ruyo committed:
不正データによるクラッシュ対応

ruyo committed:
不正データによるクラッシュ対応

ruyo committed:
optionalをpairに置き換え

ruyo committed:
SpecularTexture対応

ruyo committed:
Update README.md

ruyo committed:
ZOffsetについて、DepthとShadowに使い分けできるようにした

syui committed:
fix: typo

ruyo committed:
UE5.7Previewに対応

ruyo committed:
ビルドスクリプト更新

ruyo committed:
リリーススクリプト更新

ruyo committed:
Update README.md

ruyo committed:
Outlineマテリアルに パラメータコピー

ruyo committed:
Outlineマテリアル パラメータ適用

ruyo committed:
アウトライン補正修正

ruyo committed:
マテリアルStaticSwitchを取得できるようにした

ruyo committed:
アウトラインZOffset 判定方法を変更

ruyo committed:
プラグイン依存関係を整理

ruyo committed:
Merge pull request #544 from syui/linux-support

Fix linux build error

ruyo committed:
RimフィルタがScreenPercentageを受けないようにした

ruyo committed:
CSで太さにDepthを参照するように。

ruyo committed:
CS ステンシル追加

ruyo committed:
cs 画角補正

ruyo committed:
ポーズとMorphで同名があった時、ポーズ側を登録しないようにした

ruyo committed:
UE5.7でインポート後にモデルが開けないのを修正

ruyo committed:
fix build error 5.6

ruyo committed:
fix: build error 4.27

ruyo committed:
Springチェック

ruyo committed:
Update README.md

ruyo committed:
schema追加

ruyo committed:
バリデーションチェック 準備中

ruyo committed:
サンプルのrigのパスが変わっているのを対処

ruyo committed:
VRM Validataion 有効化

ruyo committed:
vrm validationチェック有効化、オプションで無効に変更可能

ruyo committed:
IKRIGより不要なChainを削除

ruyo committed:
fix: build error 4.27

ruyo committed:
Update README.md

ruyo committed:
マテリアル生成ツールを追加

ruyo committed:
ReplaceMaterial整頓

ruyo committed:
ValidationをPKGにも適用

ruyo committed:
プレフィクス変更対応

ruyo committed:
ReplaceMaterial カスタム対応

ruyo committed:
MaterialReplace 整頓

ruyo committed:
MatCapカスタムNormal対応

ruyo committed:
キャラマテリアル EmissiveAlpha

ruyo committed:
EmissiveAlpha修正

ruyo committed:
VRM1で不正アクセスが起きることがあるのを対処

ruyo committed:
rim 設定チェック追加

ruyo committed:
Update README.md

ruyo committed:
Glitter追加

ruyo committed:
zoffset default追加

desuqcafe committed:
docs: Add fork notice and link to wiki documentation

- Added Off World Live fork notice to README files
- Linked to wiki for extended skeleton support documentation
- Removed deprecated Docs folder (documentation moved to wiki)
- Preserved original VRM4U README content

desuqcafe commented:
Created our dedicated wiki section,
Rebased onto the latest updates,
Removed our docs / personal README
Modified the original README to include 

> **Off World Live Fork:** This is a fork of [VRM4U](https://github.com/ruyo/VRM4U) with extended support for non-VRM skeletons. See our [Wiki](https://github.com/Off-World-Live/VRM4U/wiki).

campbellwmorgan reviewed (approved):
This is great! Merging...
